site/source/docs/api_reference/vr.h.rst

```diff
@@ -0,0 +1,378 @@
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:param VREye whichEye: which eye to query parameters for.
+	:param VREyeParameters eyeParam: receives the parameters for requested eye.
+	:returns: |display-function-return-doc|
+	:rtype: bool
+
+Render Loop
+===========
+
+In contrast to the usual emscripten main loop (see :ref:`emscripten-h-browser-execution-environment`),
+VR displays require their own rendering loop which is independent from the main loop. The rendering
+loop can be set per display and will act like a main loop with timing mode ``EM_TIMING_RAF`` until the
+display is requested to present, as of which it will run at the VR display's refresh rate.
+
+.. c:function:: void emscripten_vr_set_display_render_loop(VRDisplayHandle handle, em_vr_callback_func callback)
+
+	Set a C function as the per frame rendering callback of a VR display.
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|: id of the display to set the render loop for.
+	:param em_vr_callback_func callback: C function to set as per frame rendering callback.
+	:rtype: |display-function-return-doc|
+
+.. tip:: There can be only *one* render loop function per VR display. To change the render loop function, first :c:func:`cancel <emscripten_vr_cancel_display_render_loop>` the current loop, and then call this function to set another.
+
+.. c:function:: void emscripten_vr_set_display_render_loop_arg(VRDisplayHandle handle, em_vr_callback_func callback, void* arg)
+
+	Set a C function as the per frame rendering callback of a VR display.
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:param em_vr_callback_arg_func callback: C function to set as per frame rendering callback. The function signature must have a ``void*`` parameter for passing the ``arg`` value.
+	:param void* arg: User-defined data passed to the render loop function, untouched by the API itself.
+	:rtype: |display-function-return-doc|
+
+.. c:function:: void emscripten_vr_cancel_display_render_loop(VRDisplayHandle handle: |display-handle-parameter-doc|)
+
+	Cancels the render loop of a VR display should there be one running for it.
+
+	|render-loop-info|
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:rtype: |display-function-return-doc|
+
+.. c:function:: int emscripten_vr_get_frame_data(VRDisplayHandle handle)
+
+	Get view matrix, projection matrix, timestamp and head pose for current frame.
+	Only valid when called from within a render loop callback.
+
+	|render-loop-info|
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:rtype: |display-function-return-doc|
+
+.. c:function:: int emscripten_vr_submit_frame(VRDisplayHandle handle)
+
+	Submit the current state of canvases passed via VRLayerInit to
+	:c:func:`emscripten_vr_request_present` to be rendered to the VR display.
+	Only valid when called from within a render loop callback.
```

Comment:
This is very much like regular GL/D3D present? (https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglSwapBuffers.xhtml, https://msdn.microsoft.com/en-us/library/windows/desktop/bb174576(v=vs.85).aspx) I.e. until one calls `submitFrame`, nothing is actually shown on screen?

What happens if one returns from the VR display callback without calling `emscripten_vr_submit_frame` function? I suppose the previous frame that was presented with an earlier `emscripten_vr_submit_frame` call will continue to show on the screen?

Does one need to always pair each VR display callback with exactly one `emscripten_vr_submit_frame` call? What happens if one calls that function twice in a callback? Or if one called that exactly once every second callback?

Reply:
>  I suppose the previous frame that was presented with an earlier emscripten_vr_submit_frame call will continue to show on the screen

The VR driver will probably handle this as a missed a frame and try to timewarp a new frame according to the head pose. If you do that a couple times in a row, it will either fade out of your application to some driver specific scene (which notifies the user that the app is not cooperating) or maybe just display the last frame. The spec doesn't seem to handle this.

> Does one need to always pair each VR display callback with exactly one emscripten_vr_submit_frame call?

Rather: pair every `emscripten_vr_get_frame_data` with `emscripten_vr_submit_frame`... the draft for 2.0 already changes how this works.

Is that a question out of interest or should this be clarified in the docs? In the latter case: I am unsure whether (re-)explaining the WebVR API in the emscripten docs makes sense.